From 14047f8a99fedae6ebc8d75d34004734de0cba3e Mon Sep 17 00:00:00 2001
From: Koji Matsuoka <koji.matsuoka.xm@renesas.com>
Date: Thu, 6 Sep 2018 19:54:36 +0900
Subject: [PATCH 327/909] drm: rcar-du: Fix dot clock lower limit value of LVDS

commit 7b734ddfbfb9c95d4a18136115bb8533320dae8d from
git://git.kernel.org/pub/scm/linux/kernel/git/horms/renesas-bsp.git

According to H/W manual v0.54, add the dot clock lower
limit value of LVDS in r8a77990/r8a77995. This patch adds it.

Signed-off-by: Koji Matsuoka <koji.matsuoka.xm@renesas.com>
Signed-off-by: Meng Li <Meng.Li@windriver.com>
---
 drivers/gpu/drm/rcar-du/rcar_lvds.c | 12 ++++++++++--
 1 file changed, 10 insertions(+), 2 deletions(-)

diff --git a/drivers/gpu/drm/rcar-du/rcar_lvds.c b/drivers/gpu/drm/rcar-du/rcar_lvds.c
index a34800ea1da5..da2dd449c34a 100644
--- a/drivers/gpu/drm/rcar-du/rcar_lvds.c
+++ b/drivers/gpu/drm/rcar-du/rcar_lvds.c
@@ -486,11 +486,19 @@ static bool rcar_lvds_mode_fixup(struct drm_bridge *bridge,
 				 const struct drm_display_mode *mode,
 				 struct drm_display_mode *adjusted_mode)
 {
+	struct rcar_lvds *lvds = bridge_to_rcar_lvds(bridge);
+
 	/*
 	 * The internal LVDS encoder has a restricted clock frequency operating
-	 * range (31MHz to 148.5MHz). Clamp the clock accordingly.
+	 * range (31MHz to 148.5MHz). In case of r8a77990/r8a77995, frequency
+	 * operating range (5MHz to 148.5MHz). Clamp the clock accordingly.
 	 */
-	adjusted_mode->clock = clamp(adjusted_mode->clock, 31000, 148500);
+	if (lvds->info->quirks & RCAR_LVDS_QUIRK_EXT_PLL)
+		adjusted_mode->clock = clamp(adjusted_mode->clock,
+					     5000, 148500);
+	else
+		adjusted_mode->clock = clamp(adjusted_mode->clock,
+					     31000, 148500);
 
 	return true;
 }
-- 
2.17.1

